Lib/shutil.py

```diff
@@ -213,6 +213,23 @@ def _fastcopy_sendfile(fsrc, fdst):
                 _USE_CP_SENDFILE = False
                 raise _GiveupOnFastCopy(err)
 
+            if err.errno == errno.ENODATA:
+                # In rare cases, sendfile() on Linux Lustre returns
+                # ENODATA.
+                _USE_CP_SENDFILE = False
+
+                dstpos = os.lseek(outfd, 0, os.SEEK_CUR)
+                if dstpos > 0:
+                    # Some data has already been written but we use
+                    # sendfile in a mode that does not update the
+                    # input fd position when reading. Hence seek the
+                    # input fd to the correct position before falling
+                    # back on POSIX read/write method. Since sendfile
+                    # requires mmapable infd, it should also be seekable
+                    os.lseek(infd, dstpos, os.SEEK_SET)
```

Comment:
```suggestion
                # 'infd' and 'outfd' are assumed to be seekable,
                # as they are checked to be "regular" files.
                if offset > 0:
                    # Some data has already been written but we use
                    # sendfile in a mode that does not update the
                    # input fd position when reading. Hence seek the
                    # input fd to the correct position before falling
                    # back on POSIX read/write method.
                    os.lseek(infd, offset, os.SEEK_SET)
```

AFAIK, `offset` would match `dstpos` right? (or maybe offset + 1, I didn't think about it much here)

Reply:
I will check this but I think this would not work, because offset is updated only at the bottom with the return value of sendfile() (which is the number of bytes actually sent).   If are catching the exception then no return value (I think it would be available in C). So this is why the offset is not used in in the next condition but rather the return of lseek, even if it is just to check no data have been written.

Reply:
Thanks for the comments. I think I would leave the `dstpos = os.lseek(outfd, 0, os.SEEK_CUR)` in case sendfile has failed after writing some data. This would be consistent with the usage below (`Give up on first call and if no data was copied` and subsequent line) which was I think done for same reason.